### PR TITLE
#45 comment articles

### PR DESCRIPTION
**Pull request by Banovvv**
Implemented logic to add comments to articles. Added unit tests.

### CODE REVIEW

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 1]
```diff
@@ -1,33 +1,65 @@
...
-        public async Task<IActionResult> GetByArticleAsync(int articleId)
+        public async Task<ActionResult<IEnumerable<Comment>>> GetByArticleAsync(int articleId)
         {
-            var comments = await _service.GetByArticleAsync(articleId);
+            var comments = await _commentService.GetByArticleAsync(articleId);
 
             if (!comments.Any())
             {
-                return BadRequest($"No comments for this article");
+                return BadRequest(ValidationMessages.NoCommentsForArticle);
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:37]: If there are no comments, you shouldn't return `BadRequest`.

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 2]
```diff
@@ -1,33 +1,65 @@
...
             }
 
             return Ok(comments);
         }
+
+        [Authorize]
+        [HttpPost("PostComment")]
+        public async Task<ActionResult> PostCommentAsync([FromBody] PostCommentDTO commentInput)
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:45]: Please do not use HTTP verbs in model name. Please rename `PostCommentDTO` to `CreateCommentRequest`

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 3]
```diff
@@ -1,33 +1,65 @@
...
+        {
+            var validationResult = await _commentValidator.ValidateAsync(commentInput);
+
+            if (!validationResult.IsValid)
+            {
+                var response = _generateModelStateDictionary.modelStateDictionary(validationResult);
+                return ValidationProblem(response);
+            }
+
+            var created = await _commentService.PostCommentAsync(commentInput);
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:55]: Please do not use HTTP verbs in name of service method. Please rename to  `AddComment`.

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 4]
```diff
@@ -1,33 +1,65 @@
...
+
+            if (!created)
+            {
+                return BadRequest(ValidationMessages.UnableToPostComment);
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:59]: Please do not use HTTP verbs in response messages.

File: SportsHub/SportsHub.Api/Validations/CommentValidator.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+using SportsHub.DAL.Data.Configurations.Constants;
+
+namespace SportsHub.Api.Validations
+{
+    public class CommentValidator : AbstractValidator<PostCommentDTO>
+    {
+        public CommentValidator()
+        {
+            RuleFor(x => x.Content)
+                .NotEmpty().WithMessage("Content is required")
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Validations/CommentValidator.cs:12]: Magic strings.

File: SportsHub/SportsHub.Application/Services/CommentService.cs [hunk 6]
```diff
@@ -16,5 +17,28 @@ public async Task<IEnumerable<Comment>> GetByArticleAsync(int id)
...
+            }
+
+            var comment = new Comment()
+            {
+                Content = commentInput.Content,
+                AuthorId = commentInput.AuthorId,
+                ArticleId = commentInput.ArticleId,
+            };
+
+            await _unitOfWork.CommentRepository.PostCommentAsync(comment);
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Application/Services/CommentService.cs:38]: Please do not use HTTP verbs in name of repository method.

File: SportsHub/UnitTests/Controllers/CommentControllerTests.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            var comments = CommentMockData.GetForArticle();
+            _commentService.Setup(service => service.GetByArticleAsync(acticleId)).ReturnsAsync(comments);
+
+            //Act
+            var result = await _commentController.GetByArticleAsync(acticleId);
+            var resultObject = TestHelper.GetObjectResultContent<IEnumerable<Comment>>(result);
+
+            //Assert
+            Assert.IsType<OkObjectResult>(result.Result);
+            Assert.Equal(3, resultObject.Count());
```
Comment (MBocharskyi) [inline SportsHub/UnitTests/Controllers/CommentControllerTests.cs:54]: Magic number.

File: SportsHub/UnitTests/Controllers/CommentControllerTests.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        [Fact]
+        public async Task GetByArticleAsync_CommentsForProvidedActicleDoNotExist_ReturnsBadRequest()
+        {
+            //Arrange
+            int acticleId = 5;
```
Comment (MBocharskyi) [inline SportsHub/UnitTests/Controllers/CommentControllerTests.cs:61]: Code duplication.

File: SportsHub/UnitTests/Controllers/CommentControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        public async Task GetByArticleAsync_CommentsForProvidedActicleExist_ReturnsOkStatus()
+        {
+            //Arrange
+            int acticleId = 5;
+            var comments = CommentMockData.GetForArticle();
+            _commentService.Setup(service => service.GetByArticleAsync(acticleId)).ReturnsAsync(comments);
+
+            //Act
+            var result = await _commentController.GetByArticleAsync(acticleId);
+            var resultObject = TestHelper.GetObjectResultContent<IEnumerable<Comment>>(result);
```
Comment (MBocharskyi) [inline SportsHub/UnitTests/Controllers/CommentControllerTests.cs:50]: How this line of code relates to `Act`?

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs (hunk 1 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/CommentController.cs:37]: Fixed.

File: SportsHub/UnitTests/Controllers/CommentControllerTests.cs (hunk 7 above)
Comment (Banovvv) [reply SportsHub/UnitTests/Controllers/CommentControllerTests.cs:54]: Fixed.

File: SportsHub/UnitTests/Controllers/CommentControllerTests.cs (hunk 8 above)
Comment (Banovvv) [reply SportsHub/UnitTests/Controllers/CommentControllerTests.cs:61]: Fixed.

File: SportsHub/UnitTests/Controllers/CommentControllerTests.cs (hunk 9 above)
Comment (Banovvv) [reply SportsHub/UnitTests/Controllers/CommentControllerTests.cs:50]: It does not, I've moved it further down the file.

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs (hunk 3 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/CommentController.cs:55]: Fixed.

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs (hunk 4 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/CommentController.cs]: Fixed.

File: SportsHub/SportsHub.Api/Validations/CommentValidator.cs (hunk 5 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Validations/CommentValidator.cs]: Fixed.

File: SportsHub/SportsHub.Application/Services/CommentService.cs (hunk 6 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Services/CommentService.cs:38]: Fixed.

**Review by deivit98 — APPROVED** [review]

**Comment by MBocharskyi** [comment]
Please resolve conflicts.

**Comment by Banovvv** [comment]
> Please resolve conflicts.

Done

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 10]
```diff
@@ -1,33 +1,65 @@
...
     public class CommentController : ControllerBase
     {
-        private readonly ICommentService _service;
+        private readonly ICommentService _commentService;
         private readonly IMapper _mapper;
+        private readonly IValidator<CreateCommentDTO> _commentValidator;
+        private readonly IGenerateModelStateDictionary _generateModelStateDictionary;
 
-        public CommentController(ICommentService service, IMapper mapper)
+        public CommentController(ICommentService service, IMapper mapper, IValidator<CreateCommentDTO> commentValidator, IGenerateModelStateDictionary generateModelStateDictionary)
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:22]: add new line for every parameter

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 11]
```diff
@@ -1,33 +1,65 @@
...
+            {
+                var response = _generateModelStateDictionary.modelStateDictionary(validationResult);
+                return ValidationProblem(response);
+            }
+
+            var created = await _commentService.AddCommentAsync(commentInput);
+
+            if (!created)
+            {
+                return BadRequest(ValidationMessages.UnableToAddComment);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:59]: please, use throwing an exception instead true-false. Merge last main

File: SportsHub/SportsHub.Application/Services/CommentService.cs [hunk 12]
```diff
@@ -19,5 +21,28 @@ public async Task<IEnumerable<Comment>> GetByArticleAsync(int id)
...
+            var userExists = _unitOfWork.UserRepository.GetById(commentInput.AuthorId) != null;
+            var articleExists = _unitOfWork.ArticleRepository.GetById(commentInput.ArticleId) != null;
+
+            if (!userExists || !articleExists)
+            {
+                return false;
+            }
+
+            var comment = new Comment()
+            {
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Application/Services/CommentService.cs:36]: you can do it with mapper

File: SportsHub/SportsHub.DAL/Repository/CommentRepository.cs [hunk 13]
```diff
@@ -17,5 +17,10 @@ public async Task<IEnumerable<Comment>> GetByArticleAsync(int id)
                 .Where(x => x.ArticleId == id)
                 .ToListAsync();
         }
+
+        public async Task AddCommentAsync(Comment comment)
+        {
+            await DbSet.AddAsync(comment);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Repository/CommentRepository.cs:23]: add this to the generic repository, please. I think every entity has add method

**Review by MBocharskyi — APPROVED** [review]

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs (hunk 10 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/CommentController.cs:22]: Updated.

File: SportsHub/SportsHub.Application/Services/CommentService.cs (hunk 12 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Services/CommentService.cs:36]: Updated to use `_mapper`

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs (hunk 11 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/CommentController.cs:59]: Implemented

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 14]
```diff
@@ -1,33 +1,63 @@
...
+
+            if (!validationResult.IsValid)
+            {
+                var response = _generateModelStateDictionary.modelStateDictionary(validationResult);
+                return ValidationProblem(response);
+            }
+
+            await _commentService.AddCommentAsync(commentInput);                       
+
+            return Ok(ValidationMessages.CommentAddedSuccessfully);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:60]: let's return Created if there post action

File: SportsHub/SportsHub.Api/Program.cs [hunk 15]
```diff
@@ -54,6 +54,7 @@
 builder.Services.AddMvc();
 builder.Services.AddControllers();
 builder.Services.AddValidatorsFromAssemblyContaining<UserValidator>();
+builder.Services.AddValidatorsFromAssemblyContaining<CommentValidator>();
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Program.cs:57]: is it from a different assembly than UserValidator?

Comment (Banovvv) [reply SportsHub/SportsHub.Api/Program.cs:57]: No. removed it

**Review by NataliiaBidiuk — APPROVED** [review]

**Pull request merged**